base of dockerfile frontend

tonistiigi opened the pull request:
`buildctl build --frontend dockerfile.v0 --local context=. --local dockerfile=./hack/dockerfiles --frontend-opt filename=test.Dockerfile --frontend-opt target=unit-tests`

Shares the parser code with moby/moby thanks to https://github.com/moby/moby/pull/33492 . The dockerfile frontend itself will be probably moved to a docker-org repo in the future and hopefully, a frontend can be defined with an image instead of at build time then.

Only some commands are implemented. No support for `ARG` or the commands that only modify image config without having an effect for the build. For the commands that only modify image config I found that lot of them are missing from the oci spec. So if this can't be resolved we need to create a package with docker compatible spec to fully support `docker build`.

@simonferquel @AkihiroSuda

AkihiroSuda commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -0,0 +1,206 @@
+	return nil
+}
+
+// func dispatchOnbuild(d *dispatchState, c *instructions.OnbuildCommand) error {
```
Please open GitHub issues for ONBUILD and COPY?

tonistiigi replied:
#108 #109

AkihiroSuda commented on vendor.conf:
```diff
@@ -33,7 +33,7 @@ github.com/google/shlex 6f45313302b9c56850fc17f99e40caebce98c716
 
 github.com/BurntSushi/locker 392720b78f44e9d0249fcac6c43b111b47a370b8
-github.com/docker/docker 6301ac0c27aef52a96820482a01869457ae416f7 https://github.com/mcandre/moby.git
+github.com/docker/docker typed-builder-commands-squashed https://github.com/simonferquel/docker.git
```
Please use commit hash?

AkihiroSuda commented:
The current implementation does not seem to support `FROM <stagename>`:
e.g.
```console
$ cat Dockerfile
FROM alpine:latest AS foo
RUN echo hello

FROM foo
RUN echo hello2

$ buildctl build --frontend dockerfile.v0 --local context=.
INFO[0000] tracing logs to /tmp/buildctl865317890
[+] Building 2.2s (1/1) FINISHED
 => local://context                                                                                                                0.0s
 => => transferring context: 110B                                                                                                  0.0s
buildd: failed to solve: rpc error: code = Unknown desc = server message: insufficient_scope: authorization failed
```

This means BuildKit cannot build BuiltKit itself using `hack/dockerfiles/test.Dockerfile`, but I think we can merge this PR if the issue is known.

tonistiigi commented:
@AkihiroSuda Fixed that case and added some basic testcases. Need more tests that actually check the output but need to figure out what is the best method to validate the correctness of the graph. I think if we just validated the digests, the tests would break too often.

AkihiroSuda commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -0,0 +1,229 @@
+		return &allStages[len(allStages)-1].state, nil
+	}
+
+	state, ok := stagesByName[opt.Target]
```
ToLower(opt.Target)?

AkihiroSuda commented:
LGTM but some nit

Also, unrelated to PR, but envs from the image config seems not propagated.

```
$ buildctl build --frontend dockerfile.v0 --local context=${buildkit-itself}
INFO[0000] tracing logs to /tmp/buildctl166311205
[+] Building 8.7s (8/8) FINISHED
 => local://context                                                                                                                0.0s
 => => transferring context: 1.971kB                                                                                               0.0s
 => local://context                                                                                                                0.0s
 => => transferring context: 131.3MB                                                                                               5.1s
 => docker-image://docker.io/library/alpine@sha256:1072e499f3f655a032e88542330cf75b02e7bdf673278f701d7ba61629ee3ebe                1.4s
 => => resolve docker.io/library/alpine@sha256:1072e499f3f655a032e88542330cf75b02e7bdf673278f701d7ba61629ee3ebe                    1.4s
 => docker-image://docker.io/library/golang:1.8-alpine                                                                             1.5s
 => => resolve docker.io/library/golang:1.8-alpine                                                                                 1.5s
 => CACHED /bin/sh -c apk add --no-cache g++ linux-headers                                                                         0.0s
 => CACHED /bin/sh -c apk add --no-cache git make                                                                                  0.0s
 => cp -a /src /dest                                                                                                               0.4s
 => ERROR /bin/sh -c go build -o /buildd.exe ./cmd/buildd                                                                          0.1s
------
 > /bin/sh -c go build -o /buildd.exe ./cmd/buildd:
/bin/sh: go: not found
------
buildd: failed to solve: rpc error: code = Unknown desc = worker failed running [/bin/sh -c go build -o /buildd.exe ./cmd/buildd]: exit
code 127
```

AkihiroSuda commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -0,0 +1,229 @@
+		return nil, err
+	}
+
+	_ = metaArgs
```
what is this?

tonistiigi replied:
this is "args-in-from", currently unhandled like the rest of args/variable expansion.

AkihiroSuda commented on cmd/buildctl/build.go:
```diff
@@ -37,6 +37,14 @@ var buildCommand = cli.Command{
 		},
+		cli.StringFlag{
+			Name:  "frontend",
+			Usage: "Define frontend for build result",
```
"build result" -> "generating LLB"?

tonistiigi commented:
@AkihiroSuda Updated. This error was caused by the state getting lost in the mounts for `cp` https://github.com/moby/buildkit/pull/106/files#diff-b6e1be4db0a6b8d1aada0afaaa76287dR194

Also added support for specifying filename and changed the syntax so dockerfile uses a separate local source. This works better for clients that pass dockerfile from stdin.

`buildctl build --frontend dockerfile.v0 --local context=. --local dockerfile=./hack/dockerfiles --frontend-opt filename=test.Dockerfile --frontend-opt target=unit-tests`

AkihiroSuda commented:
awesome, thank you :+1:
